datafusion/catalog/src/dynamic_file/catalog.rs

```diff
@@ -0,0 +1,250 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! dynamic_file contains [`DynamicFileCatalog`] that creates tables from file paths
+//! if the wrapped [`CatalogProviderList`] doesn't have the table provider.
+
+use crate::{CatalogProvider, CatalogProviderList, SchemaProvider, TableProvider};
+use async_trait::async_trait;
+#[cfg(feature = "home_dir")]
+use dirs::home_dir;
+use std::any::Any;
+use std::sync::Arc;
+
+/// Wrap another catalog provider list
+pub struct DynamicFileCatalog {
```

Comment:
I wonder if it would make more sense to call this DynamicFileCatalog with the idea that eventually it would support arbitrary urls, not just local files?

Reply:
I will try it tomorrow. 🤔

Reply:
After some research, I think `DynamicFile` still makes sense to me. A file could refer to either a local file (implicitly or explicitly containing the `files://` protocol) or a remote file (which requires the corresponding object store). Alternatively, `DynamicUrlCatalog` also seems appropriate. We have `UrlTableFactory` to build the `ListingTable`.